### PR TITLE
feat(rust, python): activate cse for group_by (again)

### PR DESCRIPTION
**Pull request by ritchie46**
Also made the optimization more robust. Fixed a few bugs as drive by's that were found by activating cse on all rust tests.

### CODE REVIEW

File: crates/polars-plan/src/logical_plan/optimizer/cse_expr.rs [hunk 1]
```diff
@@ -256,40 +256,45 @@ impl ExprIdentifierVisitor<'_> {
 
     /// return `None` -> node is accepted
     /// return `Some(_)` node is not accepted and apply the given recursion operation
+    /// `Some(_, true)` don't accept this node, but can be a member of a cse.
+    /// `Some(_,  false)` don't accept this node, and don't allow as a member of a cse.
     fn accept_node(&self, ae: &AExpr) -> Option<(VisitRecursion, bool)> {
+        // Don't allow this node, but allow as a member of a cse.
```
Comment (reswqa) [inline crates/polars-plan/src/logical_plan/optimizer/cse_expr.rs:262]: I guess this comment should be swapped with the comments of `REFUSE_ALLOW_MEMBER`.

Comment (ritchie46) [reply crates/polars-plan/src/logical_plan/optimizer/cse_expr.rs:286]: Good catch!

**Comment by reswqa** [comment]
I think we add an extra `cse` col at the end of `df`, but we does not update the schema with this new column. Which maybe the reason of test panic.

**Comment by ritchie46** [comment]
> I think we add an extra `cse` col at the end of `df`, but we does not update the schema with this new column. Which maybe the reason of test panicking. Of course, in non test environments, falling back to linear search does not result in errors but is inefficient

Yes, that was one part. Adapted the branch to start linear search at the location we inserted the temporary columns.

**Pull request merged**